infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/multiuser/oauth/OpenshiftTokenInitializationFilter.java

```diff
@@ -0,0 +1,141 @@
+import org.eclipse.che.multiuser.api.authentication.commons.filter.MultiUserEnvironmentInitializationFilter;
+import org.eclipse.che.multiuser.api.authentication.commons.token.RequestTokenExtractor;
+import org.eclipse.che.multiuser.api.permission.server.AuthorizedSubject;
+import org.eclipse.che.multiuser.api.permission.server.PermissionChecker;
+import org.eclipse.che.workspace.infrastructure.openshift.OpenShiftClientFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Singleton
+public class OpenshiftTokenInitializationFilter extends MultiUserEnvironmentInitializationFilter {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(OpenshiftTokenInitializationFilter.class);
+
+  private static final List<String> UNAUTHORIZED_ENDPOINT_PATHS =
+      Collections.singletonList("/system/state");
+
+  private final PermissionChecker permissionChecker;
+  private final OpenShiftClientFactory clientFactory;
+
+  private final UserManager userManager;
+
+  @Inject
+  public OpenshiftTokenInitializationFilter(
+      SessionStore sessionStore,
+      RequestTokenExtractor tokenExtractor,
+      OpenShiftClientFactory clientFactory,
+      UserManager userManager,
+      PermissionChecker permissionChecker) {
+    super(sessionStore, tokenExtractor);
+    this.clientFactory = clientFactory;
+    this.userManager = userManager;
+    this.permissionChecker = permissionChecker;
+  }
+
+  @Override
+  protected String getUserId(String token) {
+    try {
+      return getCurrentUser(token).getMetadata().getUid();
+    } catch (KubernetesClientException e) {
+      if (e.getCode() == 401) {
+        LOG.error(
+            "Unauthorized when getting current user. Invalid OpenShift token, probably expired. Re-login? Re-request the token?");
+      }
+      throw new RuntimeException(e);
+    } catch (InfrastructureException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  protected Subject extractSubject(String token) {
+    try {
+      ObjectMeta userMeta = getCurrentUser(token).getMetadata();
+      User user =
+          userManager.getOrCreateUser(
+              userMeta.getUid(), openshiftUserEmail(userMeta), userMeta.getName());
+      return new AuthorizedSubject(
+          new SubjectImpl(user.getName(), user.getId(), token, false), permissionChecker);
+    } catch (InfrastructureException | ServerException | ConflictException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private io.fabric8.openshift.api.model.User getCurrentUser(String token)
+      throws InfrastructureException {
+    OpenShiftClient client = clientFactory.createAuthenticatedOC(token);
+    return client.currentUser();
+  }
+
+  protected String openshiftUserEmail(ObjectMeta userMeta) {
+    // OpenShift User does not have data about user's email. However, we need some email. For now,
+    // we can use fake email, but probably we will need to find better solution.
```

Comment:
it's time to think about removing a mandatory flag from email. wdyt?

Reply:
I'm not aware of any usages. I'll try to remove it.

Reply:
That is something TBD. doesn't have to do immediate actions.

Reply:
Yes, I would keep it separate from this PR. If you're ok with that, I would use here fake email. However, we have to do it before moving new auth out of experimetal into supported features.

Reply:
ok